### ISSUE TITLE
Extensions registered within test class hierarchy are ordered bottom-up instead of top-down

### ISSUE THREAD

**Issue by sbrannen** [junit-team/junit5#208]
While testing support for using `@ExtendWith` within a test class hierarchy in conjunction with #207, it became apparent that extensions will currently be registered bottom-up instead of top-down, with the latter being the expected behavior.

## Example

Given the following use case with extensions `A` and `B`, one would expect extension `A` to be registered before `B` for test class `ExtendedTests`; however, the opposite is true.

````java
	@ExtendWith(A.class)
	class BaseTests {
	}

	@ExtendWith(B.class)
	class ExtendedTests extends BaseTests {
	}
````

## Related Tests

See the `findRepeatableAnnotationsWithMultipleAnnotationsOnSuperclass` and `findRepeatableAnnotationsWithContainerAnnotationOnSuperclass` test cases in `AnnotationUtilsTests` for concrete examples.

**Comment by jlink** [comment]
@sbrannen It's definitely a bug that should be fixed.

But it also makes me think that if we have to rely on the order of registration something in our extension mechanism could be improved to make order of registration less relevant. The order should only matter if two extensions compete about the order of an extension point being applied.

In the example from SpringExtension that triggered the issue: Could the problem be circumvented by choosing an appropriate `Position` value when registering the conflicting extension points in the extensions?

**Comment by jlink** [comment]
After revisiting the current implementation (https://github.com/junit-team/junit5/blob/master/junit-commons/src/main/java/org/junit/gen5/commons/util/AnnotationUtils.java#L144-L229) I'm not so sure anymore, we should fix it. The algorithm is already complicated enough. Adding the behaviour suggested in the description would add even more complexity and just open the can of other corner cases, for example:

````java
	@ExtendWith(A.class)
	class BaseTests {
	}

	@ExtendWith(B.class)
	@ExtendWith(A.class)
	class ExtendedTests extends BaseTests {
	}
````

What's more intuitive: `A.class` being registered before `B.class` or after `B.class`? Consider that the author of `ExtendedTests` might not be aware of `BaseTests` already having `A.class` as extension - or it is added after the fact.

In order to fix the underlying issue, extensions shouldn't rely on the order of registration. My hope is that with the approach in #135 most of these problems could be solved by adding approriate values for `Position`. Maybe we will need a different kind of position constraints between extensions, like `Position.before(A.class, B.class)`. And now we're back in the discussion about #112.

So my vote is for not fixing this and instead concentrate on eventually come to grips with #135 and #112.

**Comment by marcphilipp** [comment]
@sbrannen Has this already been addressed in f77221e?

**Comment by marcphilipp** [comment]
Was fixed in f77221e.
